tsl/src/remote/connection.c

```diff
@@ -170,6 +171,147 @@ typedef struct TSConnection
  */
 static ListNode connections = { &connections, &connections };
 
+static bool
+fill_simple_error(TSConnectionError *err, int errcode, const char *errmsg, const TSConnection *conn)
+{
+	if (NULL == err)
+		return false;
+
+	MemSet(err, 0, sizeof(*err));
+
+	err->errcode = errcode;
+	err->msg = errmsg;
+	err->host = pstrdup(PQhost(conn->pg_conn));
+	err->nodename = pstrdup(NameStr(conn->node_name));
+
+	return false;
+}
+
+static bool
+fill_connection_error(TSConnectionError *err, int errcode, const char *errmsg,
+					  const TSConnection *conn)
+{
+	if (NULL == err)
+		return false;
+
+	fill_simple_error(err, errcode, errmsg, conn);
+	err->connmsg = pstrdup(PQerrorMessage(conn->pg_conn));
+
+	return false;
+}
+
+static char *
+get_error_field_copy(const PGresult *res, int fieldcode)
+{
+	const char *msg = PQresultErrorField(res, fieldcode);
+
+	if (NULL == msg)
+		return NULL;
+	return pchomp(msg);
+}
+
+/*
+ * Convert libpq error severity to local error level.
+ */
+static int
+severity_to_elevel(const char *severity)
+{
+	/* According to https://www.postgresql.org/docs/current/libpq-exec.html,
+	 * libpq only returns the severity levels listed below. */
+	static const struct
+	{
+		const char *severity;
+		int elevel;
+	} severity_levels[] = { {
+								.severity = "ERROR",
+								.elevel = ERROR,
+							},
+							{
+								.severity = "FATAL",
+								.elevel = FATAL,
+							},
+							{
+								.severity = "PANIC",
+								.elevel = PANIC,
+							},
+							{
+								.severity = "WARNING",
+								.elevel = WARNING,
+							},
+							{
+								.severity = "NOTICE",
+								.elevel = NOTICE,
+							},
+							{
+								.severity = "DEBUG",
+								.elevel = DEBUG1,
+							},
+							{
+								.severity = "INFO",
+								.elevel = INFO,
+							},
+							{
+								.severity = "LOG",
+								.elevel = LOG,
+							},
+							/* End marker */
+							{
+								.severity = NULL,
+								.elevel = 0,
+							} };
+	int i;
+
+	if (NULL == severity)
+		return 0;
+
+	i = 0;
+
+	while (NULL != severity_levels[i].severity)
+	{
+		if (strcmp(severity_levels[i].severity, severity) == 0)
+			return severity_levels[i].elevel;
+		i++;
+	}
+
+	pg_unreachable();
+
+	return ERROR;
+}
+
+/*
+ * Fill a connection error based on the result of a remote query.
+ */
+static bool
+fill_result_error(TSConnectionError *err, int errcode, const char *errmsg, const PGresult *res)
```

Comment:
This always returns `false`: you can remove the return type (or is the last `return false` a typo?)

Reply:
Yes, that's intentional since it allows returning from other functions in one-line instead of two, like so:

```
if (error)
   return fill_error(....);
```

instead of:

``` 
if (error)
{
      fill_error(...);
      return false;
)
```

If you feel strongly that we should always do the second option, I can change.

Reply:
No strong opinion, this is not important and you can leave it as it is, but piggybacking that functionality on the function seems to be unnecessary. To stick it on one line one could just as well write:
```
if (error)
  return fill_error(...), false;
```